cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/types/PointerType.kt

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2019, Fraunhofer AISEC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.graph.types
+
+import de.fraunhofer.aisec.cpg.graph.Name
+import java.util.*
+import org.neo4j.ogm.annotation.Relationship
+
+/**
+ * PointerTypes represent all references to other Types. For C/CPP this includes pointers, as well
+ * as arrays, since technically arrays are pointers. For JAVA the only use case are arrays as there
+ * is no such pointer concept.
+ */
+class PointerType : Type, SecondOrderType {
+    @Relationship(value = "ELEMENT_TYPE") override lateinit var elementType: Type
+
+    enum class PointerOrigin {
+        POINTER,
+        ARRAY
+    }
+
+    var pointerOrigin: PointerOrigin? = null
+        private set
+
+    private constructor()
+    constructor(elementType: Type, pointerOrigin: PointerOrigin?) : super() {
+        language = elementType.language
+        if (pointerOrigin == PointerOrigin.ARRAY) {
+            name = elementType.name.append("[]")
+        } else {
+            name = elementType.name.append("*")
+        }
+        this.pointerOrigin = pointerOrigin
+        this.elementType = elementType
+    }
+
+    constructor(type: Type?, elementType: Type, pointerOrigin: PointerOrigin?) : super(type!!) {
```

Comment:
probably needs a default constructor or one with default arguments here, for neo4j